docs/sources/tempo/metrics-from-traces/metrics-queries/sampling-guide.md

```diff
@@ -0,0 +1,950 @@
+
+- **Real-time dashboards** requiring fast refresh rates
+- **Exploratory data analysis** where approximate results accelerate insights
+- **Resource-constrained environments** with limited compute capacity
+- **Large-scale deployments** processing terabytes of trace data daily
+
+Grafana Tempo supports three complementary sampling approaches: adaptive sampling that automatically optimizes based on query characteristics, fixed span-level sampling for precise control over span selection, and trace-level sampling optimized for trace-wide aggregations.
+
+## Before you begin
+
+TraceQL metrics sampling requires:
+
+- Tempo 2.8+ with TraceQL metrics enabled
+- `local-blocks` processor configured in metrics-generator
+- Grafana 10.4+ or Grafana Cloud for UI integration
+
+## Sampling methods deep dive
+
+### Adaptive sampling: `with(sample=true)`
+
+Adaptive sampling represents the most sophisticated sampling approach, automatically determining the optimal sampling strategy based on query selectivity, data volume, and result distribution. The sampler analyzes query patterns and adjusts sampling rates dynamically to maintain accuracy while maximizing performance gains.
+
+The adaptive sampler monitors query execution in real-time, tracking the rate at which matching spans are found. When data is abundant (high match rates), it reduces the sampling percentage to maintain performance. For rare events or selective queries, it maintains higher sampling rates or disables sampling entirely to preserve signal.
+
+```traceql
+// Basic adaptive sampling for service rate metrics
+{ resource.service.name="checkout-service" } | rate() with(sample=true)
+
+// Adaptive sampling with grouping
+{ status=error } | count_over_time() by (resource.service.name) with(sample=true)
+
+// Complex query with adaptive sampling
+{ resource.service.name="api-gateway" && span.http.method="POST" }
+| quantile_over_time(duration, .95, .99) by (span.http.route) with(sample=true)
+```
+
+#### Best use cases
+
+- **Heavy aggregation queries** like `{ } | rate()` that process millions of spans
+- **Dashboard queries** where consistent performance matters more than exact precision
+- **User-driven exploration** when users need quick insights from large datasets
+- **Multi-service analysis** with unpredictable data volumes across services
+
+#### Technical characteristics
+
+- Automatically switches between span-level and trace-level sampling based on query type
+- Uses probabilistic sampling with reservoir sampling for quantile calculations
+- Applies back pressure when downstream processing becomes a bottleneck
+- Scales sampling factors based on actual data encountered, not estimated volumes
+
+#### Limitations
+
+- May over-sample rare events, reducing performance benefits for needle-in-haystack queries
+- Introduces slight non-determinism in results across repeated executions
```

Comment:
Huh I wonder where this came from. It's not true, the sampling is actually deterministic based on the block so it would be the same across runs of the same query on the same block.

Reply:
That's weird.